src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/RequestBodyReader.cs

```diff
@@ -0,0 +1,181 @@
+                        bytesRead = await messageBody.ReadAsync(writableBuffer.Buffer.GetArray(), cancellationToken);
+                        writableBuffer.Advance(bytesRead);
+                    }
+                    finally
+                    {
+                        writableBuffer.Commit();
+                    }
+
+                    if (bytesRead == 0)
+                    {
+                        _pipe.Writer.Complete();
+                        return;
+                    }
+
+                    await writableBuffer.FlushAsync();
+                }
+            }
+            catch (Exception ex)
+            {
+                _pipe.Writer.Complete(ex);
+            }
+        }
+
+        public void Reset()
+        {
+            _pipe.Reader.Complete();
+            _pipe.Writer.Complete();
+            _pipe.Reset();
+        }
+
+        public virtual async Task<int> ReadAsync(ArraySegment<byte> buffer, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            while (true)
+            {
+                var result = await _pipe.Reader.ReadAsync();
+                var readableBuffer = result.Buffer;
+
+                try
+                {
+                    if (!readableBuffer.IsEmpty)
+                    {
+                        var count = Math.Min(result.Buffer.Length, buffer.Count);
+                        readableBuffer = result.Buffer.Slice(0, count);
+                        readableBuffer.CopyTo(buffer);
+                        return count;
+                    }
+                    else if (result.IsCompleted)
+                    {
+                        return 0;
+                    }
+                }
+                finally
+                {
+                    _pipe.Reader.Advance(readableBuffer.End);
+                }
+            }
+        }
+
+        public virtual async Task CopyToAsync(Stream destination, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            while (true)
+            {
+                var result = await _pipe.Reader.ReadAsync();
+                var readableBuffer = result.Buffer;
+
+                try
+                {
+                    if (!readableBuffer.IsEmpty)
+                    {
+                        foreach (var memory in readableBuffer)
+                        {
+                            var array = memory.GetArray();
+                            await destination.WriteAsync(array.Array, array.Offset, array.Count);
+                        }
+                    }
+                    else if (result.IsCompleted)
+                    {
+                        return;
+                    }
+                }
+                finally
+                {
+                    _pipe.Reader.Advance(readableBuffer.End);
+                }
+            }
+        }
+
+        public Task ConsumeAsync(CancellationToken cancellationToken = default(CancellationToken))
```

Comment:
Seemed to help a little when I was fixing perf. And we already had something like that in `MessageBody`, so why not?